### DISCUSSION TITLE
Parameter bypass flag

### DISCUSSION THREAD

**Discussion by TaleTN** [free-audio/clap#156] [General]
Hi,

I'm adding CLAP support to my own IPlug1 fork, and I'm looking into adding support for CLAP_PARAM_IS_BYPASS. However, its documentation seems somewhat lacking (IMHO), so I have a lot of questions:

1. Is the host required to support this flag? I think it probably should, but I guess it's too late for that now.
2. Should the host hide the parameter? I think it shouldn't, but it would seem that some hosts already do this.
3. Can/should the parameter be automatable? I don't think so, because this might conflict with the host's own bypass if that's also automatable.
4. Who can change this parameter, only the host, or also the plugin?
5. Who is responsible for loading/saving the state of this parameter, the host or the plugin?
6. What should be the initial value be? I.e. can I make a plugin that is bypassed by default?
7. Wouldn't it be better if there were a simple extension that provided this functionality? That way it can be optional on both the host and plugin side, without it messing with the plugin parameter list or other side-effects.

**Comment by jaxter184** [comment]
I think it's perfectly reasonable for the answers to many of these questions to be up to the host.
* 1,2,3: I could see these going either way depending on how the host works. Nothing should really break in the interface between the plugin and the host regardless of what decisions the host makes.
* 4,5,6: I think just treating it as a regular parameter is fine. I don't see any benefit in giving bypass a special case.
* 7: I could see there being an extension, but I think the way it is now is more flexible and easier for both hosts and plugins to support.

**Reply by TaleTN** [reply]
Thanks for your reply.

Treating it as a regular parameter seems like a good starting point. However, the problem is that both Bitwig and REAPER already don't seem to do this. Bitwig e,.g. doesn't display the parameter in its parameter list, but it does display its current value when you change it, though without the parameter name. And REAPER hides the parameter from the genertic UI, though it does include it for automation, but it doesn't seem to allow me to change the parameter from inside my plugin GUI.

For most our plugins we already have a regular, automatable VST2/AU parameter that controls bypass, but it would seem that enabling the CLAP_PARAM_IS_BYPASS flag on this parameter causes more problems than it solves, so as it is I guess we won't be supporting this flag then (for now anyway).

**Comment by leslie-sanford** [comment]
I'm struggling to implement a bypass parameter with the CLAP_PARAM_IS_BYPASS flag. I've gotten it to work in Reaper. When bypassing the plugin using its own UI, Reaper recognizes this and bypasses the plugin at the host level. The reverse is true. If you bypass the plugin in the host, it sends a parameter event corresponding to the bypass parameter so that the plugin can update its state.

I've not been able to get this to work in Bitwig. I could be missing something, but the necessary two-way communication just doesn't seem to be happening.
